Visual editor for `.quilt/catalog/config.yaml`

fiskus opened the pull request:
![buckets-preferences-editor](https://github.com/user-attachments/assets/1944c9fa-c182-4481-bff5-920d89f00556)

<!-- Remove items that are irrelevant to this PR -->

- [x] Unit tests
- [x] Documentation
- [x] [Changelog](../tree/master/docs/CHANGELOG.md) entry (skip if change is not significant to end users, e.g. docs only)

nl0 commented on catalog/app/components/FileEditor/QuiltConfigEditor/BucketPreferences/State.tsx:
```diff
@@ -0,0 +1,200 @@
+  ext: Partial<Defaults>,
+): KeyedValue<K> {
+  return {
+    isDefault: !user[key],
```
this means unchecked values are always considered "default" -- is this intended?

nl0 commented on catalog/app/components/FileEditor/QuiltConfigEditor/BucketPreferences/BucketPreferences.tsx:
```diff
@@ -0,0 +1,358 @@
+      fullWidth
+      size={size}
+    >
+      <M.InputLabel>Default source bucket</M.InputLabel>
```
```suggestion
      <M.InputLabel>Default bucket</M.InputLabel>
```

"source" is implied by the context

fiskus commented on catalog/app/components/FileEditor/QuiltConfigEditor/BucketPreferences/State.tsx:
```diff
@@ -0,0 +1,201 @@
+
+type Defaults = Required<ReturnType<typeof parseUser>>
+
+const sys: Defaults = {
```
I would like to replace `utils/BucketPreferences` with this implementation instead. Though, I don't want to do it in this PR. It seems, that parsing implementation in this PR is cleaner and more straightforward, however it not that clean that I planned to make.

nl0 replied:
ok, then a FIXME or XXX comment would do

nl0 commented on catalog/app/components/FileEditor/QuiltConfigEditor/BucketPreferences/BucketPreferences.tsx:
```diff
@@ -0,0 +1,347 @@
+}
+
+const I18N = {
+  'ui.actions': 'Toggle buttons visibility',
```
> I wanted to emphasize that these checkboxes just show or hide buttons, and they don't enable or disable functionality behind these buttons.

maybe a single paragraph explaining this in a single place somewhere above the checkboxes would work?

fiskus commented on catalog/app/components/FileEditor/QuiltConfigEditor/BucketPreferences/BucketPreferences.tsx:
```diff
@@ -0,0 +1,347 @@
+  'ui.blocks.browser': 'Toggle files listings on bucket and packages pages',
+  'ui.blocks.code': 'Toggle CODE section',
+  'ui.blocks.meta': 'Toggle METADATA section',
+  'ui.blocks.qurator': 'Enable Qurator omni',
```
I think it has attributes from both actions and sections and should be in its own section like Athena.

fiskus replied:
I'd skip "context dependent", because it confuses me without further explanation

nl0 replied:
> I think it has attributes from both actions and sections and should be in its own section like Athena.

nope, it currently only affects the visibility of the little "assistant" buttons. qurator-related sections were removed

fiskus replied:
The "assistant" buttons open the Qurator section in the same way the "CODE" wide button opens the Code section (or "ANALYTICS" button opens the Analytics section). If we move the Code section to a popup triggered by a button, we would still retain the toggle in "ui.blocks.code," wouldn't we? The actions within "ui.actions" are focused on server-side operations, such as package creation or downloading/removing objects.

In contrast, the Qurator section is not strictly tied to modifying server-side elements. Its primary purpose often seems to be assisting with package analysis.

Given this, it seems reasonable to consider that "qurator" doesn't fully align with the scope of "ui.actions" or "ui.blocks." Instead, it stands out as its own distinct feature, with the potential to support additional functionalities unrelated to other actions or sections.

nl0 replied:
> The "assistant" buttons open the Qurator section in the same way the "CODE" wide button opens the Code section (or "ANALYTICS" button opens the Analytics section).

there's no "qurator" sections anymore, they were removed prior to stable release of qurator omni

> The actions within "ui.actions" are focused on server-side operations, such as package creation or downloading/removing objects.

i didn't think about it that way. in my model, an "action" does something, and a "block" shows something.
qurator option was introduced to hide qurator sections, but later was repurposed to hide qurator buttons without changing the underlying key names.

> it stands out as its own distinct feature, with the potential to support additional functionalities unrelated to other actions or sections.

maybe, but currently it doesn't seem to be worth it, imo, since it aligns just fine with a common-sense definition of an "action"

fiskus replied:
> there's no "qurator" sections anymore

No horizontal section similar to Code and Analytics. But there is a vertical Qurator section, isn't it?

> an "action" does something, and a "block" shows something

I agree. You can easily say that "Copy package" copies the package. The result of that is a new package. But what does Qurator do? And what is a result of this action? It rather shows sophisticated responses and make an alternative TUI for the web. 

Anyway, I don't mind if you change the "ui.blocks.qurator" to "ui.actions.qurator" (though, "ui.qurator" would be better), but I don't know how to do this gracefully, because someone can use this feature already. Probably, first we need to deprecate one toggle, and support both.

nl0 replied:
> No horizontal section similar to Code and Analytics. But there is a vertical Qurator section, isn't it?

well, kinda, but this is a very different thing, bc it's not a part of the bucket UI and it isn't affected by that flag (it's always accessible via a floating button)

> But what does Qurator do? And what is a result of this action? It rather shows sophisticated responses and make an alternative TUI for the web.

this is not about what qurator does in general, but what the button does -- it invokes qurator with a pre-filled message (e.g. "summarize this document")

> Anyway, I don't mind if you change the "ui.blocks.qurator" to "ui.actions.qurator" (though, "ui.qurator" would be better), but I don't know how to do this gracefully, because someone can use this feature already. Probably, first we need to deprecate one toggle, and support both.

i'm ok with just keeping it as it is in the model for now (for compatibility reasons -- don't want to deal with it now), as long as we hide behind a ui that makes sense.
and in my opinion it makes sense to display it under the "actions" group along with the package creation and stuff, bc it's a button same as the others, invoking an action

fiskus replied:
I moved qurator to its own section in this editor

nl0 commented on catalog/app/components/FileEditor/QuiltConfigEditor/BucketPreferences/PackageDescription.tsx:
```diff
@@ -0,0 +1,277 @@
+      onBlur={() => onChange(value)}
+      onChange={handleChange}
+      onKeyDown={handleEnter}
+      placeholder="namespace/production"
```
```suggestion
      placeholder="e.g. ^prefix/.*"
```
looking at the code, i see that regexp matching is used in every case, there's no literal matching, so `namespace/production` will match every package name containing that substring, not only that specific package. for literal matching, a pattern must be anchored with `^` and `$`.
we should fix the docs btw, because the example there is incorrect

nl0 commented on catalog/app/components/FileEditor/QuiltConfigEditor/BucketPreferences/PackageDescription.tsx:
```diff
@@ -0,0 +1,277 @@
+          error={!!error}
+          helperText={error?.message}
+          label="JSON paths from `user_meta`"
+          placeholder="$.Some.Key, $.Another.Key"
```
```suggestion
          placeholder="e.g. $.Some.Key, $.Another.Key"
```

fiskus commented on catalog/app/components/FileEditor/QuiltConfigEditor/BucketPreferences/BucketPreferences.tsx:
```diff
@@ -0,0 +1,471 @@
+  'ui.blocks.qurator': 'Enable Qurator assistance',
+
+  'ui.blocks.meta.*.expanded': {
+    title: 'Metadata in the package list',
```
Yes, I confused it with `packageDescription`

fiskus commented on catalog/app/components/FileEditor/QuiltConfigEditor/BucketPreferences/BucketPreferences.tsx:
```diff
@@ -0,0 +1,471 @@
+  'ui.blocks.qurator': 'Qurator assistance',
+
+  'ui.blocks.meta.*.expanded': {
+    title: 'Metadata in the package list',
```
```suggestion
    title: 'Metadata on the package page',
```

fiskus commented on catalog/app/components/FileEditor/QuiltConfigEditor/BucketPreferences/BucketPreferences.tsx:
```diff
@@ -0,0 +1,471 @@
+  if (key === 'ui.package_description.multiline') {
+    return 'ui.package_description.all'
+  }
+  if (key === 'ui.blocks.qurator') {
```
I'll rename it so, I hope, it would be more clear and less confusing, that these keys are not config keys but keys for groups that are not strictly matched with config.

fiskus replied:
`ui.blocks.qurator`  → 'custom_group.qurator'
`ui.package_description.all` → `custom_group.common_package_description`

fiskus commented on catalog/app/components/FileEditor/QuiltConfigEditor/BucketPreferences/BucketPreferences.tsx:
```diff
@@ -0,0 +1,471 @@
+  'ui.actions.revisePackage': '"REVISE PACKAGE" on the package page',
+  'ui.actions.writeFile': 'Buttons to create or edit files',
+
+  'ui.qurator': 'Qurator',
```
I renamed `ui.qurator` to `custom_group.qurator`

nl0 reviewed (commented):
1. let's switch common and selective package list settings, since imo it makes sense to show common settings first;
2. i still feel like you should bring back the borders around package group display settings, or use any other means to visually group the controls (e.g. one group per line or smth);
3. pls add a placeholder for athena workgroup input -- it looks somewhat weird when empty;
4. let's move qurator flag to the actions section, bc it's confusing to have it on its own, given the flag toggles the display of very specific buttons (let's label it `"Qurator Assistance" icon button on object and package pages` for clarity)

fiskus commented:
> let's switch common and selective package list settings, since imo it makes sense to show common settings first

I agree, but then we need to re-phrase the content. Because, the description throws some `user_meta` in common section, and only then introduce it in a more specific section. So, description flow became broken

nl0 commented:
> > let's switch common and selective package list settings, since imo it makes sense to show common settings first
> 
> I agree, but then we need to re-phrase the content. Because, the description throws some `user_meta` in common section, and only then introduce it in a more specific section. So, description flow became broken

ok, your call.
a suggestion: merge these two groups, show common settings first (a single flag currently), then a description about the selective settings application, and then the actual groups

fiskus commented:
Oh, 3 of 4 :) I didn't try moving qurator to actions. But I feel, I need to rewrite a significant portion of the code for that.

nl0 reviewed (approved):
> I didn't try moving qurator to actions. But I feel, I need to rewrite a significant portion of the code for that.

really? is it that hard to move a setting to another group?

> My opinion: your last 4 propositions make sense, but they are not very significant, on the other hand they increase complexity and total amount of work.

well, they seem quite significant from where i stand, since these issues affect the overall **quality** of the product, making the UX sloppier and somewhat more confusing, but whatever.

PS: please don't forget to update the screenshot and crush it (i use pngquant for this e.g. `pngquant --quality=80-100 --force --skip-if-larger --strip -o imgs/admin-tabulator-settings.png imgs/admin-tabulator-settings.png`)

fiskus commented:
> is it that hard to move a setting to another group?

UI is constructed programmatically from config values. If I need to move one value from one place to another, I need to make a first and single exception. I also can rename the field in the config during parsing, but it opens another Pandora's box I didn't consider yet.

fiskus commented:
* moved `ui.blocks.qurator` to Actions, added borders for package descriptions and moved common package settings above selective (also changed the description)
* fixed problematic keys: `sourceBuckets`, `package_description_multiline`, `defaultSourceBucket`
* fixed empty YAML issue
* fixed handling change in Metadata fields for package descriptions
* wrote tests for cases where I made mistakes

nl0 commented on catalog/app/components/FileEditor/QuiltConfigEditor/BucketPreferences/BucketPreferences.tsx:
```diff
@@ -421,17 +409,18 @@ const useStyles = M.makeStyles((t) => ({
-// Group keys so 'a.b.c', and 'a.b.d' are in the same 'a.b' group
-// Also, move some config keys into standalone groups
+// Group keys so 'ui.a.b', and 'ui.a.b' are in the same 'ui.a' group
+// Inclose single root keys into one-field groups
```
```suggestion
// Enclose single root keys into one-field groups
```

fiskus commented:
Thanks. Updated screenshot
